File: src/leetCodeProblems/TwoDArrayMatrix/SprialOrderMatrixI54.java
package leetCodeProblems.TwoDArrayMatrix;

/**
 * LeetCode - https://leetcode.com/problems/spiral-matrix/
 * 
 */
import java.util.*;

public class SprialOrderMatrixI54 {
	public List<Integer> spiralOrder(int[][] matrix) {
        
        List<Integer> output = new ArrayList<Integer>();
        
        if (matrix.length == 0) {
            return output;
        }
        
        // Clockwise direction traversal
        int[] directionX = {0, 1, 0,-1};
        int[] directionY = {1, 0, -1, 0};
        
        int rows = matrix.length;
        int columns = matrix[0].length;
        
        boolean[][] visited = new boolean[rows][columns]; 
        
        int direction = 0;
        
        int currentRowIndex = 0;
        int currentColumnIndex = 0;
        
        for(int i=0; i < rows*columns; i++) {
            
            output.add(matrix[currentRowIndex][currentColumnIndex]);
            visited[currentRowIndex][currentColumnIndex] = true;
            
            int nextRowIndex = currentRowIndex + directionX[direction];
            int nextColumnIndex = currentColumnIndex + directionY[direction];
            
            // This is positive case.
            if ( 0 <= nextRowIndex &&
               nextRowIndex < rows &&
                 0 <= nextColumnIndex &&
               nextColumnIndex < columns &&
               !visited[nextRowIndex][nextColumnIndex]) {
                
                currentRowIndex = nextRowIndex;
                currentColumnIndex = nextColumnIndex;
                
            }
            else { // Otherwise change direction
                
                direction = (direction + 1 ) % 4; // This is IMPORTANT code.
                
                currentRowIndex = currentRowIndex + directionX[direction];
                currentColumnIndex = currentColumnIndex + directionY[direction];
                
            }  
        }
        
        return output;
    }
}

File: src/leetCodeProblems/Matrix/SprialOrderMatrixI54.go
package main

/*
- LeetCode - https://leetcode.com/problems/spiral-matrix/
- Time - O(m*n)
- Space - O(m*n)
*/
func spiralOrder(matrix [][]int) []int {
	dX := []int{0, 1, 0, -1}
	dY := []int{1, 0, -1, 0}

	direction := 0
	curRow := 0
	curCol := 0

	out := []int{}
	m := make([][]bool, len(matrix))

	for i := 0; i < len(matrix); i++ {
		m[i] = make([]bool, len(matrix[0]))
	}

	for i := 0; i < len(matrix)*len(matrix[0]); i++ {
		out = append(out, matrix[curRow][curCol])
		m[curRow][curCol] = true

		nextRow := curRow + dX[direction]
		nextCol := curCol + dY[direction]

		if nextRow >= 0 && nextRow < len(matrix) &&
			nextCol >= 0 && nextCol < len(matrix[0]) &&
			!m[nextRow][nextCol] {
			curRow = nextRow
			curCol = nextCol
		} else {
			direction = (direction + 1) % 4

			curRow += dX[direction]
			curCol += dY[direction]
		}
	}
	return out
}
